### ISSUE TITLE
Add WriteInPublic to pa.org.za

### ISSUE THREAD

**Issue by paullenz** [mysociety/pombola#1682]

**Comment by mhl** [comment]
Probably the simplest thing to do is to create a new nightly Popolo export that only includes the politicians you'd want to be able to write to, then have a new PopIt instance that imports that, as we do with the current PopIt mirrors. The WriteInPublic could be based on that.

Then the question is whether this needs to appear to be embedded in the site or it's OK for it to look like a separate service. (The former would need more work.)

The same comments apply to https://github.com/mysociety/pombola/issues/1480

**Comment by paullenz** [comment]
For ease, I would suggest a separate service that is styled to look like
the main pa site and linked to from it

**Comment by JenMysoc** [comment]
From PMG: At this stage in our democracy, it would be a step too far to include WARD COUNCILLORS. At most, National Assembly MPs and possibly NCOP and MPLs [should be included in this tool]

**Comment by JenMysoc** [comment]
From PMG: During our first 2017 Skype conversation with you, we talked about adding the Write to Them feature this year to each MP's profile. Can we assign this to the next sprint? Following the latest political dramas in SA, there has been an increased call for people’s  ability to be able to reach out to their MPs. During the Skype conversation we talked about the being able to moderated  the comments made to the MP.

This is now a high priority for them and if we could discuss when we might be able to look at this it would be very useful! It would be great to capitalise on the increased interest in contacting MPs :)

**Comment by JenMysoc** [comment]
To do this I think we need to:

- [ ] Set up an instance of WiP including the ZA parliament and their emails
- [ ] Skin the instance of WiP with something so it looks like pa.org.za
- [ ] Create an admin account or allow multiple admin accounts so that people from PMG can moderate messages and send them on
- [ ] Import messages from Mp message pages back into PA

**Comment by tmtmtmtm** [comment]
Assuming the WiP is going to be backed by EP, then we also need to fix some of the issues that are making it very difficult to keep EP in sync with PA, most importantly an equivalent of what we did for Mzalendo in https://github.com/mysociety/pombola/pull/2172

**Comment by JenMysoc** [comment]
This has been paired down and now PMG want the following;

A WriteIt API instance that allows people to send emails to the central email address for a committee, some of these emails should either be pre-formatted or have simple questions in a form like element to enable ease of creating a message for people who aren't sure what to say. The form should be on the homepage or similar.

They want this to look or be similar to the form here http://www.news24.com/SouthAfrica/Politics/no-confidence-vote-email-baleka-mbete-20170720

**Comment by JenMysoc** [comment]
Ok this has been scaled back up. Now what is needed is:

1 x WiP Instance to write to MPs which does the following
- write in public tool to write to NA MPs 
- the tool should be free text - for people to enter whatever they like
- the tool should be moderated so you can approve messages before they are sent on
- the tool should be embedded on MP pages, so people can write to MPs from their page and also see the correspondence that's been sent publicly
- they also want a link to the tool on the home page so people can go straight to a place where they can choose an MP (probably from a drop down list) and write to them

and 1 x Write to a committee tool which should;
- allow people to write free text to committees on PA
- each committee should be written to using just one central email address
- replies, and the tool, should be shown on the committee page on PA 
- The tool to write should be clickable from the homepage

**Comment by mhl** [comment]
I think that we can break down the tasks needed for this now into the following. Some of these already have tickets, but it would be a good idea to create tickets for any of the non-trivial actions (i.e. most of them) and link to those tickets from this checklist.

- [ ] Prepare JSON data for both instances
    - [ ] Prepare JSON data for the National Assembly
        - [ ] Update the EveryPolitician scraper https://github.com/everypolitician-scrapers/south-africa-national-assembly/blob/master/scraper.rb to include the Pombola person ID as a new identifier (by looking `<meta name="pombola-person-id" content="xxxx">` in each person page
        - [ ] Add a cron job to PA to fetch the EP JSON for the national assembly and update the identifiers on each person to include the EP UUID. (We can then use that to find the messages for just that person when it comes to adding a "public correspondence" tab to the person pages.)
        - There may be other changes we need to make to improve the quality of the EP list of National Assembly, but the above would certainly be helpful.
    - [ ] Prepare JSON data for committee secretaries
        -  [ ] Issue here: https://github.com/mysociety/pombola/issues/2324
- [ ] Allow filtering of messages fetched from the WriteInPublic API based on the `popolo:person` identifier: https://github.com/mysociety/alpaca/issues/70
- [ ] Add code to [send a message via WriteInPublic](https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#sending-messages) from People's Assembly. (I'd suggest that we have a new Django application to encapsulate the interactions from Pombola to WriteInPublic.)
- [ ] Update the templates for both WriteInPublic instances, as describe here: https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#templates
- [ ] Add a webhook handling endpoint to the new WriteInPublic Django application (see https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#receiving-messages)
    -  [ ] Send emails (from an @pa.org.za address) to notify message senders of any new reply
- [ ] Add a tab to the person page of anyone in the National Assembly showing any correspondence via WriteInPublic: https://github.com/mysociety/pombola/issues/2301
- [ ] Add an index page listing committees, linking through to all correspondence sent to them
- [ ] Updates of the front page to make it obvious how to write to either a National Assembly member of a Committee (via its secretary): https://github.com/mysociety/pombola/issues/2321 (That includes deciding how that interaction should work.)

**Comment by mhl** [comment]
I think that we can break down the tasks needed for this now into the following. Some of these already have tickets, but it would be a good idea to create tickets for any of the non-trivial actions (i.e. most of them) and link to those tickets from this checklist.

- [ ] Prepare JSON data for both instances
    - [ ] Prepare JSON data for the National Assembly
        - [ ] Update the EveryPolitician scraper https://github.com/everypolitician-scrapers/south-africa-national-assembly/blob/master/scraper.rb to include the Pombola person ID as a new identifier (by looking `<meta name="pombola-person-id" content="xxxx">` in each person page
        - [ ] Add a cron job to PA to fetch the EP JSON for the national assembly and update the identifiers on each person to include the EP UUID. (We can then use that to find the messages for just that person when it comes to adding a "public correspondence" tab to the person pages.)
        - There may be other changes we need to make to improve the quality of the EP list of National Assembly, but the above would certainly be helpful.
    - [ ] Prepare JSON data for committee secretaries
        -  [ ] Issue here: https://github.com/mysociety/pombola/issues/2324
- [ ] Allow filtering of messages fetched from the WriteInPublic API based on the `popolo:person` identifier: https://github.com/mysociety/alpaca/issues/70
- [ ] Add code to [send a message via WriteInPublic](https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#sending-messages) from People's Assembly. (I'd suggest that we have a new Django application to encapsulate the interactions from Pombola to WriteInPublic.)
- [ ] Update the templates for both WriteInPublic instances, as describe here: https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#templates
- [ ] Add a webhook handling endpoint to the new WriteInPublic Django application (see https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#receiving-messages)
    -  [ ] Send emails (from an @pa.org.za address) to notify message senders of any new reply
- [ ] Add a tab to the person page of anyone in the National Assembly showing any correspondence via WriteInPublic: https://github.com/mysociety/pombola/issues/2301
- [ ] Add an index page listing committees, linking through to all correspondence sent to them
- [ ] Updates of the front page to make it obvious how to write to either a National Assembly member of a Committee (via its secretary): https://github.com/mysociety/pombola/issues/2321 (That includes deciding how that interaction should work.)

**Comment by mhl** [comment]
I think that we can break down the tasks needed for this now into the following. Some of these already have tickets, but it would be a good idea to create tickets for any of the non-trivial actions (i.e. most of them) and link to those tickets from this checklist.

- [ ] Prepare JSON data for both instances
    - [ ] Prepare JSON data for the National Assembly
        - [ ] Update the EveryPolitician scraper https://github.com/everypolitician-scrapers/south-africa-national-assembly/blob/master/scraper.rb to include the Pombola person ID as a new identifier (by looking `<meta name="pombola-person-id" content="xxxx">` in each person page
        - [ ] Add a cron job to PA to fetch the EP JSON for the national assembly and update the identifiers on each person to include the EP UUID. (We can then use that to find the messages for just that person when it comes to adding a "public correspondence" tab to the person pages.)
        - There may be other changes we need to make to improve the quality of the EP list of National Assembly, but the above would certainly be helpful.
    - [ ] Prepare JSON data for committee secretaries
        -  [ ] Issue here: https://github.com/mysociety/pombola/issues/2324
- [ ] Allow filtering of messages fetched from the WriteInPublic API based on the `popolo:person` identifier: https://github.com/mysociety/alpaca/issues/70
- [ ] Add code to [send a message via WriteInPublic](https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#sending-messages) from People's Assembly. (I'd suggest that we have a new Django application to encapsulate the interactions from Pombola to WriteInPublic.)
- [ ] Update the templates for both WriteInPublic instances, as describe here: https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#templates
- [ ] Add a webhook handling endpoint to the new WriteInPublic Django application (see https://github.com/mysociety/write-it/blob/master/INTEGRATION_GUIDE.md#receiving-messages)
    -  [ ] Send emails (from an @pa.org.za address) to notify message senders of any new reply
- [ ] Add a tab to the person page of anyone in the National Assembly showing any correspondence via WriteInPublic: https://github.com/mysociety/pombola/issues/2301
- [ ] Add an index page listing committees, linking through to all correspondence sent to them
- [ ] Updates of the front page to make it obvious how to write to either a National Assembly member of a Committee (via its secretary): https://github.com/mysociety/pombola/issues/2321 (That includes deciding how that interaction should work.)

**Comment by chrismytton** [comment]
Some thoughts/questions/concerns on this now I've done a bit of investigation/prototyping:

1. How can we handle email confirmations? By default API-created messages don’t have confirmation enabled, we can turn that on but it will mean the user will end up on WiP after confirming their address
2. We’re going to need to find a way to represent committees in the Popolo that WiP is expecting. I think that will mean generating some custom Popolo to feed into WiP, but haven’t given it too much thought yet
3. What are people’s thoughts on storing a copy of messages and responses in the Pombola DB vs just getting them from the WiP API on each request?

**Comment by chrismytton** [comment]
We discussed the points in my previous comment in Slack, summary below.

> How can we handle email confirmations? By default API-created messages don’t have confirmation enabled, we can turn that on but it will mean the user will end up on WiP after confirming their address

We've decided that having email confirmations turned on and then the user ending up on WriteInPublic (styled to look vaguely like pa.org.za) would be an acceptable route to take.

> We’re going to need to find a way to represent committees in the Popolo that WiP is expecting. I think that will mean generating some custom Popolo to feed into WiP, but haven’t given it too much thought yet

This is already being tracked in #2324. Thanks @mhl for reminding me about that ticket!

> What are people’s thoughts on storing a copy of messages and responses in the Pombola DB vs just getting them from the WiP API on each request?

After some discussion we decided that initially it would be simpler to just go via the API. If that proves problematic then it shouldn't be too much of a jump to start storing things in the DB at some point down the road, but simplicity wins at this stage.

**Comment by chrismytton** [comment]
Quick update on where we're at with this.

We've got the initial version of writing to representatives working in https://github.com/mysociety/pombola/pull/2329. That's currently deployed to the staging site and has been sent over to PMG for feedback, though they're on their summer recess, so we're not expecting feedback until the new year.

Once we've got feedback from PMG and addressed anything that comes up I'd like to get #2329 merged and live and then do the next chunk of work in a slightly less monolithic pull request 😨 

Writing to committees is still outstanding. There are a few parts to this that need to be considered:

- Adding committee Popolo JSON output to Pombola for feeding into WriteInPublic - I've done a prototype version of this but needs cleaning up
- Making the datasource prefix configurable. Currently this is hardcoded to the prefix for the representatives Popolo, but needs to be different depending on where the people come from at the WriteInPublic end.
  - This might mean adding a model to Pombola that represents a WriteIt datasource which can then be associated with a Person/Organization to show what the prefix for them should be, but I need to give this a bit more thought
- Deciding on how the writing to committees bits should be done in the UI, i.e. do we want to have them in the same list as the representatives, or should it be an entirely different journey? (I'd say have it all in the same screen, as it's simpler to implement, but this needs to be discussed with a designer/PMG)
